File: python/0095.py
from typing import List, Optional
from data_strcutres import TreeNode


class Solution:
     def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def build_trees(lst):
            if len(lst) == 0:
                return [None]
            out = []
            for i in range(len(lst)):
                for left in build_trees(lst[:i]):
                     for right in build_trees(lst[i+1:]):
                         root = TreeNode(lst[i])
                         root.left = left
                         root.right = right
                         out.append(root)
            return out
        return build_trees(list(range(1, n+1)))

File: cpp/0095.cpp
#include <vector>
#include <iostream>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return build_subtrees(1, n);
    }
private:
    vector<TreeNode*> build_subtrees(int beg, int end) {
        vector<TreeNode*> lst;
        if (beg > end) {
            lst.push_back(nullptr);
            return lst;
        }
        for (int i = beg; i <= end; i++) {
            for (auto left: build_subtrees(beg, i-1)) {
                for (auto right: build_subtrees(i+1, end)) {
                    TreeNode* root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    lst.push_back(root);
                }
            }
        }
        return lst;
    }
};
